File: src/aap/schemas/alignment_card.py
"""Alignment Card schema — Agent alignment declaration.

Defines the Alignment Card structure per SPEC Section 4. An Alignment Card
is a structured document declaring an agent's alignment posture:
- Principal relationship
- Declared values
- Autonomy envelope
- Audit commitment

See SPEC.md Section 4 for complete specification.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class PrincipalType(str, Enum):
    """Type of principal the agent serves."""

    HUMAN = "human"
    ORGANIZATION = "organization"
    AGENT = "agent"
    UNSPECIFIED = "unspecified"


class RelationshipType(str, Enum):
    """Nature of authority delegation from principal to agent."""

    DELEGATED_AUTHORITY = "delegated_authority"
    """Agent acts within bounds set by principal."""

    ADVISORY = "advisory"
    """Agent provides recommendations; principal makes decisions."""

    AUTONOMOUS = "autonomous"
    """Agent operates independently within declared values."""


class Principal(BaseModel):
    """Principal relationship declaration (SPEC Section 4.3)."""

    type: PrincipalType = Field(
        ..., description="Type of principal"
    )
    identifier: str | None = Field(
        None, description="Principal identifier (DID, email, org ID)"
    )
    relationship: RelationshipType = Field(
        ..., description="Nature of authority delegation"
    )
    escalation_contact: str | None = Field(
        None, description="Endpoint for escalation notifications"
    )


class HierarchyType(str, Enum):
    """How value conflicts are resolved."""

    LEXICOGRAPHIC = "lexicographic"
    """Values are ordered by priority; higher priority wins."""

    WEIGHTED = "weighted"
    """Values have weights; weighted sum determines outcome."""

    CONTEXTUAL = "contextual"
    """Resolution depends on context; no fixed rule."""


class ValueDefinition(BaseModel):
    """Definition of a custom value."""

    name: str = Field(..., description="Human-readable name")
    description: str = Field(..., description="What this value means operationally")
    priority: int = Field(
        default=0, description="Priority for lexicographic ordering (higher = more important)"
    )


class Values(BaseModel):
    """Value declarations (SPEC Section 4.4)."""

    declared: list[str] = Field(
        ..., description="List of value identifiers"
    )
    definitions: dict[str, ValueDefinition] | None = Field(
        None, description="Definitions for non-standard values"
    )
    conflicts_with: list[str] | None = Field(
        None, description="Values this agent refuses to coordinate with"
    )
    hierarchy: HierarchyType | None = Field(
        None, description="How value conflicts are resolved"
    )

    @model_validator(mode="after")
    def custom_values_must_be_defined(self) -> "Values":
        """Non-standard values must have definitions."""
        standard_values = {
            "principal_benefit", "transparency", "minimal_data",
            "harm_prevention", "honesty", "user_control", "privacy", "fairness",
        }
        if self.definitions:
            for value in self.declared:
                if value not in standard_values and value not in self.definitions:
                    raise ValueError(f"Custom value '{value}' must be defined in definitions")
        return self


class TriggerAction(str, Enum):
    """Action to take when escalation trigger matches."""

    ESCALATE = "escalate"
    """Defer decision to principal."""

    DENY = "deny"
    """Refuse to take the action."""

    LOG = "log"
    """Log the action but proceed."""


class EscalationTrigger(BaseModel):
    """Condition that triggers escalation (SPEC Section 4.5)."""

    condition: str = Field(
        ..., description="Condition expression (see SPEC Section 4.6)"
    )
    action: TriggerAction = Field(
        ..., description="Action to take when trigger matches"
    )
    reason: str = Field(
        ..., description="Human-readable explanation"
    )


class MonetaryValue(BaseModel):
    """Monetary value specification."""

    amount: float = Field(..., description="Numeric amount")
    currency: str = Field(default="USD", description="ISO 4217 currency code")


class AutonomyEnvelope(BaseModel):
    """Autonomy bounds and escalation triggers (SPEC Section 4.5)."""

    bounded_actions: list[str] = Field(
        ..., description="Actions permitted without escalation"
    )
    escalation_triggers: list[EscalationTrigger] = Field(
        ..., description="Conditions requiring escalation"
    )
    max_autonomous_value: MonetaryValue | None = Field(
        None, description="Maximum transaction value without escalation"
    )
    forbidden_actions: list[str] | None = Field(
        None, description="Actions never permitted"
    )


class StorageType(str, Enum):
    """Audit log storage type."""

    LOCAL = "local"
    REMOTE = "remote"
    DISTRIBUTED = "distributed"


class TamperEvidence(str, Enum):
    """Tamper-evidence mechanism for audit logs."""

    APPEND_ONLY = "append_only"
    SIGNED = "signed"
    MERKLE = "merkle"


class AuditStorage(BaseModel):
    """Audit log storage configuration."""

    type: StorageType = Field(..., description="Storage type")
    location: str | None = Field(None, description="Storage endpoint or location")


class AuditCommitment(BaseModel):
    """Audit trail commitments (SPEC Section 4.7)."""

    trace_format: str = Field(
        default="ap-trace-v1", description="Trace format identifier"
    )
    retention_days: int = Field(
        ..., ge=1, description="Minimum retention period in days"
    )
    storage: AuditStorage | None = Field(
        None, description="Storage configuration"
    )
    queryable: bool = Field(
        ..., description="Whether traces can be queried externally"
    )
    query_endpoint: str | None = Field(
        None, description="Endpoint for trace queries (required if queryable=true)"
    )
    tamper_evidence: TamperEvidence | None = Field(
        None, description="Tamper-evidence mechanism"
    )

    @model_validator(mode="after")
    def queryable_requires_endpoint(self) -> "AuditCommitment":
        """If queryable is true, query_endpoint is required."""
        if self.queryable and not self.query_endpoint:
            raise ValueError("query_endpoint is required when queryable is true")
        return self


class AlignmentCard(BaseModel):
    """Alignment Card — Agent alignment declaration (SPEC Section 4).

    A structured document declaring an agent's alignment posture. It MUST be
    machine-readable (JSON) and SHOULD be human-readable.

    Example:
        card = AlignmentCard(
            aap_version="0.1.0",
            card_id="ac-12345",
            agent_id="did:web:agent.example.com",
            issued_at=datetime.utcnow(),
            principal=Principal(type=PrincipalType.HUMAN, relationship=RelationshipType.DELEGATED_AUTHORITY),
            values=Values(declared=["principal_benefit", "transparency"]),
            autonomy_envelope=AutonomyEnvelope(
                bounded_actions=["search", "recommend"],
                escalation_triggers=[
                    EscalationTrigger(
                        condition='action_type == "purchase"',
                        action=TriggerAction.ESCALATE,
                        reason="Purchases require approval",
                    )
                ],
            ),
            audit_commitment=AuditCommitment(
                retention_days=90,
                queryable=True,
                query_endpoint="https://agent.example.com/api/traces",
            ),
        )
    """

    aap_version: str = Field(
        default="0.1.0", description="AAP specification version"
    )
    card_id: str = Field(
        ..., description="Unique identifier for this card (UUID or URI)"
    )
    agent_id: str = Field(
        ..., description="Identifier for the agent (DID, URL, or UUID)"
    )
    issued_at: datetime = Field(
        ..., description="When this card was issued"
    )
    expires_at: datetime | None = Field(
        None, description="When this card expires"
    )
    principal: Principal = Field(
        ..., description="Principal relationship declaration"
    )
    values: Values = Field(
        ..., description="Value declarations"
    )
    autonomy_envelope: AutonomyEnvelope = Field(
        ..., description="Autonomy bounds and escalation triggers"
    )
    audit_commitment: AuditCommitment = Field(
        ..., description="Audit trail commitments"
    )
    extensions: dict[str, Any] | None = Field(
        None, description="Protocol-specific extensions"
    )

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary suitable for JSON serialization."""
        return self.model_dump(mode="json", exclude_none=True)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "AlignmentCard":
        """Create from dictionary."""
        return cls.model_validate(data)

    def is_expired(self) -> bool:
        """Check if the card has expired."""
        if self.expires_at is None:
            return False
        return datetime.utcnow() > self.expires_at

    def has_value(self, value: str) -> bool:
        """Check if a value is declared."""
        return value in self.values.declared

    def is_action_bounded(self, action: str) -> bool:
        """Check if an action is in the bounded actions list."""
        return action in self.autonomy_envelope.bounded_actions

    def is_action_forbidden(self, action: str) -> bool:
        """Check if an action is forbidden."""
        return action in (self.autonomy_envelope.forbidden_actions or [])

File: typescript/src/schemas/alignment-card.ts
/**
 * Alignment Card schema - Agent alignment declaration.
 *
 * Defines the Alignment Card structure per SPEC Section 4. An Alignment Card
 * is a structured document declaring an agent's alignment posture.
 *
 * @see SPEC.md Section 4 for complete specification.
 */

/** Type of principal the agent serves. */
export type PrincipalType = "human" | "organization" | "agent" | "unspecified";

/** Nature of authority delegation from principal to agent. */
export type RelationshipType =
  | "delegated_authority"
  | "advisory"
  | "autonomous";

/** How value conflicts are resolved. */
export type HierarchyType = "lexicographic" | "weighted" | "contextual";

/** Action to take when escalation trigger matches. */
export type TriggerAction = "escalate" | "deny" | "log";

/** Audit log storage type. */
export type StorageType = "local" | "remote" | "distributed";

/** Tamper-evidence mechanism for audit logs. */
export type TamperEvidence = "append_only" | "signed" | "merkle";

/** Principal relationship declaration (SPEC Section 4.3). */
export interface Principal {
  /** Type of principal */
  type: PrincipalType;
  /** Principal identifier (DID, email, org ID) */
  identifier?: string | null;
  /** Nature of authority delegation */
  relationship: RelationshipType;
  /** Endpoint for escalation notifications */
  escalation_contact?: string | null;
}

/** Definition of a custom value. */
export interface ValueDefinition {
  /** Human-readable name */
  name: string;
  /** What this value means operationally */
  description: string;
  /** Priority for lexicographic ordering (higher = more important) */
  priority?: number;
}

/**
 * A declared value in parameterized form (SPEC Section 4.4, Phase-3.2+):
 * `id` is the value name, `domain`/`intensity` qualify how it applies.
 * `declared` accepts either this object form or a bare-string id.
 */
export interface ParameterizedValue {
  id: string;
  domain?: string;
  intensity?: string;
}

/** Value declarations (SPEC Section 4.4). */
export interface Values {
  /**
   * Behavioral and ethical values the agent applies in its decision-making.
   *
   * This field is actively monitored by AIP at runtime: every value listed here
   * is expected to appear in AP-Trace `values_applied` fields when it influences
   * a decision. Declaring a value the agent never applies produces verification
   * warnings and degrades trust scoring.
   *
   * **Include**: Ethical and behavioral commitments — e.g. `transparency`,
   * `honesty`, `accuracy`, `safety`, `accountability`, `helpfulness`,
   * `deliberation_before_action`. These describe HOW the agent reasons.
   *
   * **Do not include**: Role capabilities, operational principles, or job-function
   * descriptors — e.g. `fiduciary_precision`, `organizational_clarity`. These
   * describe WHAT the agent is in its role and belong in `extensions.clpi.role`
   * or other `extensions` metadata. Capability names (e.g. `read_documents`)
   * belong in `autonomy_envelope.bounded_actions`.
   *
   * Each entry is either a bare-string id (`"honesty"`) or a parameterized
   * object (`{id, domain?, intensity?}`). Use {@link declaredValueIds} to
   * compare against AP-Trace `values_applied` (which are always bare ids) —
   * NEVER `.includes()` raw, or object-form declarations silently mismatch.
   */
  declared: Array<string | ParameterizedValue>;
  /** Definitions for non-standard values */
  definitions?: Record<string, ValueDefinition> | null;
  /** Values this agent refuses to coordinate with */
  conflicts_with?: string[] | null;
  /** How value conflicts are resolved */
  hierarchy?: HierarchyType | null;
}

/** Condition that triggers escalation (SPEC Section 4.5). */
export interface EscalationTrigger {
  /** Condition expression (see SPEC Section 4.6) */
  condition: string;
  /** Action to take when trigger matches */
  action: TriggerAction;
  /** Human-readable explanation */
  reason: string;
}

/** Monetary value specification. */
export interface MonetaryValue {
  /** Numeric amount */
  amount: number;
  /** ISO 4217 currency code */
  currency?: string;
}

/** Autonomy bounds and escalation triggers (SPEC Section 4.5). */
export interface AutonomyEnvelope {
  /** Actions permitted without escalation */
  bounded_actions: string[];
  /** Conditions requiring escalation */
  escalation_triggers: EscalationTrigger[];
  /** Maximum transaction value without escalation */
  max_autonomous_value?: MonetaryValue | null;
  /** Actions never permitted */
  forbidden_actions?: string[] | null;
}

/** Audit log storage configuration. */
export interface AuditStorage {
  /** Storage type */
  type: StorageType;
  /** Storage endpoint or location */
  location?: string | null;
}

/** Audit trail commitments (SPEC Section 4.7). */
export interface AuditCommitment {
  /** Trace format identifier */
  trace_format?: string;
  /** Minimum retention period in days */
  retention_days: number;
  /** Storage configuration */
  storage?: AuditStorage | null;
  /** Whether traces can be queried externally */
  queryable: boolean;
  /** Endpoint for trace queries (required if queryable=true) */
  query_endpoint?: string | null;
  /** Tamper-evidence mechanism */
  tamper_evidence?: TamperEvidence | null;
}

/**
 * Alignment Card - Agent alignment declaration (SPEC Section 4).
 *
 * A structured document declaring an agent's alignment posture. It MUST be
 * machine-readable (JSON) and SHOULD be human-readable.
 */
export interface AlignmentCard {
  /** AAP specification version */
  aap_version?: string;
  /** Unique identifier for this card (UUID or URI) */
  card_id: string;
  /** Identifier for the agent (DID, URL, or UUID) */
  agent_id: string;
  /** When this card was issued (ISO 8601) */
  issued_at: string;
  /** When this card expires (ISO 8601) */
  expires_at?: string | null;
  /** Principal relationship declaration */
  principal: Principal;
  /** Value declarations */
  values: Values;
  /** Autonomy bounds and escalation triggers */
  autonomy_envelope: AutonomyEnvelope;
  /** Audit trail commitments */
  audit_commitment: AuditCommitment;
  /** Protocol-specific extensions */
  extensions?: Record<string, unknown> | null;
}

// Utility functions

/** Check if an alignment card has expired. */
export function isCardExpired(card: AlignmentCard): boolean {
  if (!card.expires_at) return false;
  return new Date() > new Date(card.expires_at);
}

/**
 * Normalize a `values.declared` list to bare-string value ids. Declared entries
 * may be bare strings OR parameterized objects (`{id, domain?, intensity?}`);
 * AP-Trace `values_applied` are always bare ids. Compare via this — a raw
 * `.includes()` would silently fail to match an object-form declaration against
 * a bare applied id, falsely flagging it `undeclared_value`.
 */
export function declaredValueIds(
  declared: ReadonlyArray<string | ParameterizedValue> | null | undefined,
): string[] {
  if (!Array.isArray(declared)) return [];
  const ids: string[] = [];
  for (const d of declared) {
    if (typeof d === "string") ids.push(d);
    else if (d && typeof d === "object" && typeof d.id === "string")
      ids.push(d.id);
  }
  return ids;
}

/** Check if a value is declared in the card (id-aware; handles parameterized form). */
export function hasValue(card: AlignmentCard, value: string): boolean {
  return declaredValueIds(card.values.declared).includes(value);
}

/**
 * Normalize `decision.values_applied` to bare string ids (ADR-065 #16).
 *
 * `values_applied` is TYPED `string[]`, but real producers (e.g. the Mnemom
 * observer's fallback/passthrough paths) emit parameterized OBJECTS
 * (`{id, intensity?, domain?}`) — the same shape declared values can take. The
 * undeclared_value check and drift value-usage tallies must compare ids, not
 * raw entries, else an object string-coerces to `'[object Object]'` and either
 * (a) never matches the declared id set → spurious `undeclared_value` deny, or
 * (b) pollutes value-usage maps with a bogus key. Reuses {@link declaredValueIds}
 * (identical string|object→id coercion); accepts the looser runtime shape.
 */
export function appliedValueIds(
  applied: ReadonlyArray<string | ParameterizedValue> | null | undefined,
): string[] {
  return declaredValueIds(applied);
}

/** Check if an action is in the bounded actions list. */
export function isActionBounded(card: AlignmentCard, action: string): boolean {
  return card.autonomy_envelope.bounded_actions.includes(action);
}

/** Check if an action is forbidden. */
export function isActionForbidden(
  card: AlignmentCard,
  action: string,
): boolean {
  return (card.autonomy_envelope.forbidden_actions ?? []).includes(action);
}
